pkg/stanza/fileconsumer/reader_factory.go

```diff
@@ -103,81 +105,107 @@ func (b *readerBuilder) build() (r *Reader, err error) {
 		Offset:          b.offset,
 		headerSettings:  b.headerSettings,
 		HeaderFinalized: b.headerFinalized,
+		FileAttributes:  b.fileAttributes,
 	}
 
 	if b.splitFunc != nil {
 		r.lineSplitFunc = b.splitFunc
 	} else {
 		r.lineSplitFunc, err = b.splitterFactory.Build(b.readerConfig.maxLogSize)
 		if err != nil {
-			return
+			return nil, err
 		}
 	}
 
-	enc, err := b.encodingConfig.Build()
+	r.encoding, err = b.encodingConfig.Build()
 	if err != nil {
-		return
+		return nil, err
 	}
-	r.encoding = enc
 
-	if b.file != nil {
-		r.file = b.file
-		r.SugaredLogger = b.SugaredLogger.With("path", b.file.Name())
-		r.FileAttributes, err = resolveFileAttributes(b.file.Name())
-		if err != nil {
-			b.Errorf("resolve attributes: %w", err)
-		}
-
-		// unsafeReader has the file set to nil, so don't try emending its offset.
-		if !b.fromBeginning {
-			if err = r.offsetToEnd(); err != nil {
-				return nil, err
-			}
-		}
+	if b.headerSettings == nil || b.headerFinalized {
+		r.splitFunc = r.lineSplitFunc
+		r.processFunc = b.readerConfig.emit
 	} else {
-		r.SugaredLogger = b.SugaredLogger.With("path", "uninitialized")
-		r.FileAttributes = &FileAttributes{}
-	}
-
-	if b.fp != nil {
-		r.Fingerprint = b.fp
-	} else if b.file != nil {
-		r.Fingerprint, err = b.readerFactory.newFingerprint(r.file)
+		// We are reading the header. Use the header split func
+		r.splitFunc = b.headerSettings.splitFunc
+		r.processFunc = r.consumeHeaderLine
+
+		// Create the header pipeline
+		r.headerPipelineOutput = newHeaderPipelineOutput(b.SugaredLogger)
+		r.headerPipeline, err = pipeline.Config{
+			Operators:     b.headerSettings.config.MetadataOperators,
+			DefaultOutput: r.headerPipelineOutput,
+		}.Build(b.SugaredLogger)
 		if err != nil {
-			return nil, err
+			return nil, fmt.Errorf("failed to build pipeline: %w", err)
+		}
+		if err = r.headerPipeline.Start(storage.NewNopClient()); err != nil {
+			return nil, fmt.Errorf("failed to start header pipeline: %w", err)
 		}
 	}
 
-	r.FileAttributes.HeaderAttributes = b.headerAttributes
-
-	if b.headerSettings == nil || b.headerFinalized {
-		r.splitFunc = r.lineSplitFunc
-		r.processFunc = b.readerConfig.emit
+	if b.file == nil {
```

Comment:
From what I can tell, we will only hit this case when calling `readerFactory.unsafeReader`. Just so I'm understanding correctly, everything before this check cannot be persisted, but everything after can be? The one behavior change I see in this method is that we won't be setting `b.headerAttributes` on the `Reader` anymore if `b.file == nil`, since we're returning early now. The new behavior seems correct to me, but I want to make sure I didn't miss something.

Reply:
> From what I can tell, we will only hit this case when calling readerFactory.unsafeReader.

Correct. 

> Just so I'm understanding correctly, everything before this check cannot be persisted, but everything after can be?

I think that's a good way to look at it. To phrase it a bit differently, prior this check we are initializing readers in a generic way. After this point, we are restoring some state from a previous reader that read this same file. 

Changes here are slightly tangential to the purpose of the PR, but rework was necessary in order to accommodate the generic attributes map. This function needs more refactoring but I think this PR improves it somewhat.

> The one behavior change I see in this method is that we won't be setting b.headerAttributes on the Reader anymore if b.file == nil, since we're returning early now.

I think this is more or less unchanged because we've flattened `headerAttributes` into `fileAttributes`, which we are setting at the beginning of the function.

Reply:
>To phrase it a bit differently, prior this check we are initializing readers in a generic way. After this point, we are restoring some state from a previous reader that read this same file.

Reading through it again, that's also a good way to look at it. Thanks for the insight.

>I think this is more or less unchanged because we've flattened headerAttributes into fileAttributes, which we are setting at the beginning of the function.

Great point, you're right; that was out of view when I was looking closely at this part of the diff and I missed it.